AD2S1210 updates

dlech opened the pull request:
This is working towards getting the AD2S1210 resolver driver out of staging.

Overview:
- The first few commits are getting the local version of the driver to exactly match mainline. This way new patches will apply cleanly and not have to be reworked before submitting upstream.
  - The first commit reverts some changes that were never merged upstream
  - All upstream patches since 5.15 are cherry-picked
  - One extra patch to backport `IIO_DMA_MINALIGN` (similar to some other drivers here)
- The remaining commits are fixes and improvements
  - Fixed bugs in probe function.
  - `fclkin` attribute is removed and replaced with devicetree binding for getting clock rate.
  - `control` attribute is removed and replaced with debugfs register access.
  - gpio device tree bindings are replaced with something that is more likely to be accepted upstream

More detailed explanations of the changes can be found in the commit messages.

Ideas for additional changes:
- `resolution` seems like a better name than `bits` for the attribute that gets and sets the resolution, but this would be a breaking change to users of the staging driver.
- `excitation_frequency` seems like a better name than `fexcit` for the attribute that gets and sets the excitation frequency, but this would be a breaking change to users of the staging driver.
- It would be nice if all of the threshold attributes used values in millivolts or millidegrees instead of the raw register value so that userspace doesn't have to do the conversion, but this would be a breaking change to users of the staging driver.
- Add an `enable_hysteresis` attribute for getting and setting the hysteresis flag in the control register. Do users need this feature?
- Add a `phase_lock_range` attribute to get and set the phase lock range flag in the control register. Do users need this feature?
- Make mode (and resolution?) gpios optional. It looks like there was some effort to do this in the reverted merge commits but it was a bit buggy. We need to define the supported scenarios first.
- Would it make more sense to make "sample-gpios" active low in the devicetree and invert the logic in the driver?
- We could do some more refactoring to reduce duplicate code when setting the excitation frequency.

Remaining tasks before submitting upstream:
- [x] add documentation for sysfs attributes
- [ ] more testing is needed to ensure iio channels are working for all supported configuration combinations


Issue: https://github.com/analogdevicesinc/linux/issues/2234

dlech commented:
FYI, the checkpatch CI failure is due to the cherry-picked upstream commits. New patches look like they passed.

dt_bindings_check CI failure is due to missing [dt-doc-validate](https://dev.azure.com/AnalogDevices/OpenSource/_build/results?buildId=29457&view=logs&j=250c57b8-394d-549e-b22b-b888ca92eef2&t=376ad7af-102b-5c8d-9bc6-bdbbc87e4bef&l=967) so seems like something not caused by these changes.

nunojsa commented on Documentation/devicetree/bindings/iio/resolver/adi,ad2s1210.yaml:
```diff
@@ -0,0 +1,112 @@
+            mode-gpios = <&gpio0 86 0>, <&gpio0 87 0>;
+            resolution-gpios = <&gpio0 88 0>, <&gpio0 89 0>;
+        };
+    };
```
commit title is not ok... `dt-binding: ...` Look at other examples please

dlech replied:
fixed

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -175,7 +156,10 @@ static int ad2s1210_config_read(struct ad2s1210_state *st,
+	ret = ad2s1210_set_mode(st, MOD_CONFIG);
+	if (ret < 0)
+		return ret;
+
```
unrelated change... This one and the above

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -298,10 +286,13 @@ static ssize_t ad2s1210_store_resolution(struct device *dev,
-	ad2s1210_set_resolution_pin(st);
+	ret = ad2s1210_set_resolution_gpios(st, udata);
+	if (ret < 0)
+		goto error_ret;
```
ditto

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -630,19 +627,44 @@ static int ad2s1210_setup_clocks(struct ad2s1210_state *st)
+	}
+
+	if (st->mode_gpios->ndescs != 2) {
+		dev_err(dev, "ad2s1210: requires exactly 2 mode-gpios\n");
```
why not `dev_err_probe()`?

dlech replied:
Because the error here is always `EINVAL`, it can never be `EPROBE_DEFER`. `dev_error_probe` is already used on the return value of `devm_gpiod_get_array` a few lines above.

dlech replied:
Looking into it a bit more, I see a few places that call `dev_err_probe()` with a constant error value that is not `EPROBE_DEFER`, so I will try it.

nunojsa replied:
yeah, dev_err_probe() is not only about `EPROBE_DEFER`. It makes it easier to handle some cases were you can set the error together with the message (turning 2 lines on code in one liner). More important is consistency. From what I have seen upstream, one should be consistent (either always use dev_err_probe() or not). But that might change from maintainer to maintainer.

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -571,9 +573,71 @@ static int ad2s1210_initial(struct ad2s1210_state *st)
+	mutex_lock(&st->lock);
+	ret = ad2s1210_config_write(st, reg, writeval);
+	mutex_unlock(&st->lock);
+
```
not a big deal but this is rather unusual... I would keep it simple. Meaning, no sanity check on the accessed registers whatsoever. This is a debug interface so it's up to userspace to not mess this up. Tyically, no one really cares to go with all this trouble in `debugfs_reg_access`

If you really care about this, check if the driver can be implemented with regmap and implement this in there

dlech replied:
I removed the checks for now, but as noted in https://github.com/analogdevicesinc/precision-converters-firmware/issues/2#issuecomment-1715756885, it is very easy to lock up the chip by trying to access a wrong register.

nunojsa replied:
Hmm, then maybe implementing regmap for this chip matters

dlech replied:
Yes, although I think there are a some complications for implementing regmap here. To read/write config registers in config mode, we would need to set `SPI_CS_WORD`, however, when not in config mode, we should not have this flag set in order to read the position an velocity so we would have to call `spi_setup()` before each use of the spi bus. Seems like more trouble than it is worth.

dlech replied:
I decided that it was worth the trouble after all.

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -151,22 +171,32 @@ static int ad2s1210_config_read(struct ad2s1210_state *st,
+	 */
+	if (address != AD2S1210_REG_FAULT && st->rx[1] & AD2S1210_ADDRESS_DATA)
+		return -EBADMSG;
+
```
as said, the neater change would be to use regmap. But I'm not sure that cs change between transfers will be doable with regmap unless you implement your own bus.

dlech replied:
done

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -634,6 +594,40 @@ static const struct iio_info ad2s1210_info = {
+	if (ret < 0)
+		return ret;
+
+	ret = devm_add_action(dev, (void(*)(void *))clk_disable_unprepare, clk);
```
nope. I see what you're doing but that is rather hacky... please just add a simple callback. But actually, when upstreaming these changes, just use `devm_clk_get_enabled()`

dlech replied:
Fixed. To minimize the diff from mainline I defined a `devm_clk_get_enabled()` locally just for the 5.15 branch.

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -634,6 +594,40 @@ static const struct iio_info ad2s1210_info = {
+	st->fclkin = clk_get_rate(clk);
+	if (st->fclkin < AD2S1210_MIN_CLKIN ||
+	    st->fclkin > AD2S1210_MAX_CLKIN) {
+		dev_err(dev, "ad2s1210: clock frequency out of range: %lu\n",
```
`dev_err_probe()`. Be consistent... dev_err_probe() is, nowadays, preferred so you might need a patch to first convert all the probe error paths to use it. Then, just keep using it when needed (like in here)

dlech replied:
Same as above, we have already checked for `EPROBE_DEFER` before this, so it can't be the case here.

dlech replied:
fixed

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -677,22 +674,29 @@ static int ad2s1210_probe(struct spi_device *spi)
-	ad2s1210_initial(st);
-
-	return 0;
+	return devm_iio_device_register(&spi->dev, indio_dev);
```
I'm not really sure if staging works like normal flow. But, if so, this patch would be one the first patch that touches the driver (fixes come first because of backporting - not sure if we really backport staging fixes). Also, a Fixes tag might be needed

dlech replied:
Ack. This is the first patch in the PR that isn't in 6.6 mainline already, so will be the first in the series submitted upstream.

nunojsa commented on drivers/staging/iio/resolver/ad2s1210.c:
```diff
@@ -4,6 +4,12 @@
+/* 5.10 compatibility */
+#include <linux/slab.h>
+#define IIO_DMA_MINALIGN		ARCH_KMALLOC_MINALIGN
+/* end 5.10 compatibility */
```
ok for now. We won't be needing this on 6.1. Also remember to drop the patch when upstreaming these changes

dlech replied:
ack.

nunojsa reviewed (commented):
Looks fairly good to me. It needs some re-arrangement of the commits (some of them are also including some multiple changes - not sure if the maintainer will like it much) but I guess you can start the upstreaming process...

Some things that I would like to mention:
* I see the bindings CI job failed (need to see if there's anything we need to update in there). So, please locally make sure that the bindings are ok.
* Other things that I saw that you can potentially clean are: remove `of_match_ptr()`, apparently no need for `spi_set_drvdata()` and `spi_setup()` is also usually not needed (called already by the spi core)

dlech commented:
> (some of them are also including some multiple changes - not sure if the maintainer will like it much)

Agreed. My first attempt was 50 patches which seemed like to much the other direction. So I will try to find some balance in between.

dlech commented:
Updates:
- Added new patch with sysfs documentation.
- Added new patch with devicetree for zed+eval board.
- Fixed device tree bindings patch subject.
- Fixed order of gpios in tables describing A0/A1/RES0/RES1 in dt-bindings.
- Added dependence on COMMON_CLK in Kconfig.
- Addressed other feedback comments as indicated above.

dlech commented:
FYI for reviewers, per https://github.com/analogdevicesinc/linux/issues/2234#issuecomment-1717273409 there will be a bit more work here before ready for review again.

dlech commented:
Updates:
- added patches to refactor and rename excitation frequency attribute and implementation
- added patch to enable getting/setting hysteresis in the control register

dlech commented:
Updates:
- changed clock name from "fclkin" to "clkin"
- Added adi,resolution device tree property
- New patch to implement phase lock range attributes
- A couple new small fix patches to fix issues in the existing mainline driver
- New patch to support the channel scale attribute

dlech commented:
Updates:
- use regmap for register access
- new patch for triggered buffer support

dlech commented:
I have rebased on the current master and added the remaining patches that were accepted upstream. 

To summarize:

- The first comment reverts all ADI tree changes to get the driver back exactly equal to the mainline staging driver as it was in upstream v6.1.
- Then all but the last commit have been cherry-picked from upstream.
  - The first 6 were made between 6.2 and 6.5.
  - Most of the remaining ones were the updates done by BayLibre.
  - This is also picking up 2 patches that introduced `linux/cleanup.h` since that is used in some of the later ad2s1210 patches.
- The final patch is adding the device tree for the AD2S1210 Eval board connected to a ZedBoard.

Note: the CI failures (DCO and checkpatch) are caused by some of the cherry-picked upstream commits, so should be safe to ignore.

Also did a quick smoke test to make sure this is still compiling and is reading data on our ZedBoard.

nunojsa commented:
> This is also picking up 2 patches that introduced linux/cleanup.h since that is used in some of the later ad2s1210 patches.

How much are you using these new macros? Normally I'm not too keen in backporting things like this but bah, this should also not give me much pain when updating to a newer kernel...

dlech commented:
> How much are you using these new macros? Normally I'm not too keen in backporting things like this but bah, this should also not give me much pain when updating to a newer kernel...

This is the only place so far. But the cleanup.h patches applied cleanly so back-porting it should be less trouble than reworking the last few patches here.
